### ISSUE TITLE
CanvasRenderingContext2D.FillText() maxWidth parameter default value doesn't work

### ISSUE THREAD

**Issue by gogglespisano** [bridgedotnet/Bridge#411]
Html5/Interfaces/CanvasRenderingContext2D.cs 

    public virtual void FillText(string text, Any<uint, int> x, Any<uint, int> y, Any<uint?, int?> maxWidth = null)

Should actually have an overload function, one with and one without maxWidth.  The default maxWidth value of null causes the text to not be displayed.

    public virtual void FillText(string text, Any<uint, int> x, Any<uint, int> y, Any<uint?, int?>)
    public virtual void FillText(string text, Any<uint, int> x, Any<uint, int> y, Any<uint?, int?> maxWidth)

**Comment by fabriciomurta** [comment]
Hello gogglespisano!

Thanks for reporting this! When I was writing support I added default values hoping Bridge wouldn't emit them as 'null' when not specified. That might have been the case by then (not really sure now) but fact is that now it really sends 'null' as the non-specified argument.

I believe, instead of sending null, optional values might have an option not to be emitted when not specified in code (thus assuming framework's supported language). But yes, the quick way around is making the overloads.

As a support, the code below:
```csharp
            int? width = 25;
            var canvas = new CanvasRenderingContext2D();
            canvas.FillText("test", 5, 5);
            canvas.FillText("test2", 15, 15, width);
```
results in
```javascript
            var width = 25;
            var canvas = new CanvasRenderingContext2D();
            canvas.fillText("test", 5, 5, null);
            canvas.fillText("test2", 15, 15, width);
```

And the 'null' argument on the third line has both a desired and undesired side effect:
- Desired: we defined on C# that default value should be null, so null is sent.
- Undesired: we didn't write anything in C# code so why emitted code has a value?
The two items above are contradictory between themselves. You fix one, you break the other.

You left a typo on your suggested overloads, they might be, rather:
```csharp
public virtual void FillText(string text, Any<uint, int> x, Any<uint, int> y);
public virtual void FillText(string text, Any<uint, int> x, Any<uint, int> y, Any<uint?, int?> maxWidth);
```

This same change must be done to several other definitions on `CanvasRenderingContext2D`.

**Comment by fabriciomurta** [comment]
Hello, I've just found (by chance) that issue #292 proposes a fix which side effect would be to fix this issue without any overloading.

**Comment by gogglespisano** [comment]
That fixes how parameter default values should be handled in generated code.  In this case, the function does not have a default value.  It truly is an overload of the function *without* the parameter.

The generated code needs to be a match to the existing HMTL5 function.  How does the fix in #292 provide that?  If I read it correctly, it proposes that the generated code produces additional code to handle an unspecified value.

**Comment by fabriciomurta** [comment]
Yes. But as the Canvas functions are not actually emitted (they are just stubbed with `[Ignore]` attribute), that code added inside the function will never be emitted, so browser's default will be used.

If it is still not clear to you how it fixes, just let me know and I'd elaborate you a more detailed explanation.

**Comment by gogglespisano** [comment]
I guess i'll need the elaboration.  How will it know that a default parameter "maxWidth = null" means don't emit the maxWidth parameter versus when someone actually wants the maxWidth value set to null?  You'd have to make sure the default value is never a value that might be used.

Also, overloading will give the proper intellisense behavior that corresponds to the actual canvas function, rather than the developer having to remember that for this function "= null" means the parameter won't be emitted.

**Comment by fabriciomurta** [comment]
## Bridge, as it is now.
```csharp
public void Test(int a, int b = 0) {
 Console.Log("Value is: "+b);
};

public void Main() {
 Test(1);
 Test(1, 2);
}
```
We will have emitted something like:
```javascript
    test: function (a, b) {
        console.log("Value is: " + b);
    },
    main: function () {
        this.test(1, 0);
        this.test(1, 2);
    }
```

Now, suppose the Test function is implemented by a library or browser/native. So we don't need to emit it, and then we'll have:
```csharp
[Ignore]
public void TestLib(int a, int b = 0) {
 Console.Log("Value is: "+b);
};

public void Main() {
 TestLib(1);
 TestLib(1, 2);
}
```
We will have emitted something like:
```javascript
    main: function () {
        this.testLib(1, 0);
        this.testLib(1, 2);
    }
```

## How Bridge would be
In the proposed solution on #292, this is how (likely) the code above will look like:
```csharp
public void Test(int a, int b = 0) {
 Console.Log("Value is: "+b);
};

public void Main() {
 Test(1);
 Test(1, 2);
}
```
We will have emitted something like:
```javascript
    test: function (a, b) {
        if (b === "undefined") b = 0;
        console.log("Value is: " + b);
    },
    main: function () {
        this.test(1);
        this.test(1, 2);
    }
```

Now, suppose the Test function is implemented by a library or browser/native. So we don't need to emit it, and then we'll have:
```csharp
[Ignore]
public void TestLib(int a, int b = 0) {
 Console.Log("Value is: "+b);
};

public void Main() {
 TestLib(1);
 TestLib(1, 2);
}
```
We will have emitted something like:
```javascript
    main: function () {
        this.testLib(1);
        this.testLib(1, 2);
    }
```

So, as you can see, the added `if (b === "undefined") b =0;` will never be emitted in the case we are adding an optional parameter for something expected to be already defined somewhere else.

> Also, overloading will give the proper intellisense behavior that corresponds to the actual canvas function, rather than the developer having to remember that for this function "= null" means the parameter won't be emitted.
Well, now this is another matter. We have two sides of the coin:
1. Doubling definition and documentation of functions with overloads on source code and diverting from original documentation
2. Improving user experience

During the writing of the framework support we emphasized following the official documentation ([here for fillText](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fillText)), so it was defined with optional arguments as documented.

So that's more like a matter of taste whether it should be an overload or show the arguments as in the documentation.

Although I agree that the 'null' display might be misleading in some situations, it is also clear that the value if not specified will not be used and that the issue in #292 is an undesired side effect in not just once situation.

I hope the explanations are clearer now. In the end, it is a matter of preferences, some will prefer overload, some would prefer to keep the code short or more concise to the official documentation. Two conflicting opinions.

**Comment by gogglespisano** [comment]
Ok, I can see how the correct code would be emitted, but I disagree with how you're interpreting the spec.

It says the parameter is "optional" and "if provided".

The code variations would be:
fillText(text, x, y);
fillText(text, x, y, maxWidth);

That is exactly what overloaded methods look like.  The spec does not say to provide null or any other value if the optional parameter is not desired.  It says to not provide the optional parameter.

It may still be somewhat of a preference, but I argue that overloads are the correct equivalent.

What would you do if the optional parameter was a string and null was an acceptable value?  How would you know whether to emit null or not emit anything?  What would be the default parameter if null is a valid value for the function?

**Comment by fabriciomurta** [comment]
Thanks for sharing your opinion, you have good points and position. Let's allow other people on the community share their opinions in the matter, though.

**Comment by erik-kallen** [comment]
Bridge 2 has an [OmitUnspecifiedArgumentsFrom] attributes that can be used to fix this scenario.

**Comment by gogglespisano** [comment]
Hi Erik, that's good to know, but the question is whether a parameter default value or a method overload is the more appropriate signature to match the canvas javascript function optional parameter?

I don't see the OmitUnspecifiedArgumentsFrom attribute in the current source.  Is Bridge 2 in a separate repository or did I miss it?

**Comment by erik-kallen** [comment]
It is currently in a non-public repo, but it will be opened up once it is reasonably ready.

As for the question regarding how to do it, my opinion is that the parameters should be included in the parameter list, except for interop with pure JS libs.

**Comment by gogglespisano** [comment]
Do you agree or disagree with Saltarelle's handling of FillText() which generates an assembly with overloads for the optional parameter?

        public void FillText( string text, double x, double y );
        public void FillText( string text, double x, double y, double maxWidth );

**Comment by erik-kallen** [comment]
Since I am the author of Saltarelle (and that is what is being used as a base for Bridge2), it would be quite strange if I didn't agree with it :)

But a declaration such as ```[OmitUnspecifiedArgumentsFrom(3)] public void FillText( string text, double x, double y, double maxWidth );``` is sometimes preferrable, especially when it comes to code declared in C# which might be invoked from either other C#, or from JS (eg. an implementation of ```IPromise.Then(Delegate fulfilledHandler, Delegate errorHandler = null, Delegate progressHandler = null)```

**Comment by DaniilVeriga** [comment]
In my opinion, we have to define overloads (which is actually done in other Bridge things like WebGL). Even if the issue with optional parameters is fixed, it will not handle the following case. For example, currently we have this method definition in CanvasRenderingContext2D:

```
public virtual void PutImageData(ImageData imagedata, int dx, int dy, Any<uint?, int?> dirtyX = null,
                                    Any<uint?, int?> dirtyY = null, Any<uint?, int?> dirtyWidth = null,
                                    Any<uint?, int?> dirtyHeight = null)
{
    return;
}
```
According to the [docs](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData) there are only two possible calls:
```
void ctx.putImageData(imagedata, dx, dy);
void ctx.putImageData(imagedata, dx, dy, dirtyX, dirtyY, dirtyWidth, dirtyHeight);
```

But with the current method definition it is possible to do this in Bridge C#:
```
ctx.putImage(imageData, dx, dy, dirtyX);
ctx.putImage(imageData, dx, dy, dirtyX, dirtyY);
ctx.putImage(imageData, dx, dy, dirtyX, dirtyY, dirtyWidth);
```

I don't think these calls are in CanvasRenderingContext2D API and are going to work.

I think that quite proves the need to define overload methods. I am working on overloads in the CanvasRenderingContext2D class.

**Comment by DaniilVeriga** [comment]
It has been fixed by adding a `.FillText()` overload method (and for other methods as well). The fix will be included into the 1.9 release.

Thank you again for reporting that!

**Comment by gogglespisano** [comment]
Thanks!
